gateway/gateway_manager.go

```diff
@@ -0,0 +1,136 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/spacemeshos/smutil/log"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/keepalive"
+)
+
+const DefaultConnTimeout = 10 * time.Second
+
+type ConnectingErrors struct {
+	errors []error
+}
+
+func (e *ConnectingErrors) Error() string {
+	// concatenate errors
+	err := ""
+	for _, e := range e.errors {
+		if err == "" {
+			err = e.Error()
+		} else {
+			err = fmt.Sprintf("%v | %v", err, e)
+		}
+	}
+	return err
+}
+
+func (e *ConnectingErrors) Is(target error) bool {
+	_, ok := target.(*ConnectingErrors)
+	return ok
+}
+
+// Manager aggregates GRPC connections to gateways.
+// Its Close() must be called when the connections are no longer needed.
+type Manager struct {
+	connections []*grpc.ClientConn
+}
+
+func (m *Manager) Connections() []*grpc.ClientConn {
+	if m == nil {
+		return []*grpc.ClientConn{}
+	}
+	return m.connections
```

Comment:
Is there any use case where the check for `nil` comes up? I would assume that if NewManager returns a non-`nil` `error` I cannot use the returned Manager. So we don't have to be more careful than necessary.

Also a caller of this function can modify the connections of the manager. If you want to prevent that you have to return a copy:

```suggestion
	con := make([]*grpc.ClientConn, len(m.connections))
	copy(con, m.connections)
	return con
```

Reply:
The check for `nil` was removed from this method, but not from the `Close` method. A caller of `Connections` can still modify the state of `Manager`.

Reply:
Yes, `nil` check is still in `Close()`, on purpose.

Regarding returning a pointer to connections, unfortunately, Go doesn't have a notion of _pointer to const_ and I don't want to return a copy.

Reply:
The `nil`-check wouldn't be necessary if we wouldn't use `defer` to call `Close` and then possibly set the value of it to `nil` before calling it 😉